### DISCUSSION TITLE
Using AD with constitutive models

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#17339] [Q&A General] [answered]
Hi:

I would like to place a set of constitutive models into a file so that they can be called by multiple ADMaterial objects.  Is it possible to propagate AD info back from these functions to the ADMaterial?  If so, what should they inherit from?  Can they just be utility functions?

For example, an ADMaterial calls a function "getHTC(Re, Pr)" where it is declared:

ADReal getHTC(ADReal Re, ADReal Pr);

Would something like that work?  Is there a good example to follow?

Thanks,
  Joe Kelly

**Comment by GiudGiud** [comment, marked as answer]
Hi Joe

This would work, the derivative information stored in the Reynolds and Prandtl number will be propagated in the function (provided ADReals are used inside too) to the output. 

There are good examples in pronghorn. For example in materials/PebbleBedKappaSolid.C 

For the inheritance question, I think maybe the thing to do is to have all these utility functions be member functions of a class, and have the ADMaterial you are creating inherit from that class. Pronhorn does that for all the closures that are similar and sharing utility functions. 
Just separate utility functions can also work. 

Best,
Guillaume

**Reply by joe61vette** [reply]
Thanks Guillaume.

I thought that might be the case but some text in the AD documentation confused me.  Namely:

 "Leveraging the automatic differentiation capabilities in MOOSE is as simple as inheriting from our AD base classes, e.g., ADKernel, ADIntegratedBC....".  

As there isn't an ADUserObject base class, I wasn't sure if AD would work with methods that inherit from UO's.

Thanks,
  Joe
